community: cleanup code, reduce complexity, add make targets

dhiller opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:

Reduces complexity in:

* [x] `validators/cmd/sigs-validator`
* [x] `generators/cmd/devstats`

Adds make targets:
* `lint` - runs golangci-lint, additionally gocyclo with `> 10`
* `test` - runs go tests
* `coverage` - creates a test coverage report

/kind cleanup
/wg code-quality

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

aburdenthehand commented:
/cc @EdDev for wg-code-quality

xpivarc commented on validators/cmd/sigs/sigs-validator_test.go:
```diff
@@ -0,0 +1,462 @@
+	"testing"
+)
+
+func Test_validateGroups(t *testing.T) {
```
Is the `_` convention? I would expect to see `TestValidateGroups`

dhiller replied:
Hm, good q - I believe that was an autogenerated test func name that I didn't really look at.

dhiller replied:
@xpivarc I investigated a bit - reason for this weird name is that Goland has auto-generated the func name for the non-exported func that is tested.

Now if I rename the func as you requested the IDE will not find the tested func anymore.

Now my q is: does it matter that much? I'd rather expect the test func name to be as close to the name of the func it tests. WDYT?

dhiller replied:
I agree that for _exported_ functions it should be as you said, however for non-exported functions whose name is starting with a lowercase letter the usage of an underscore might be more obvious.

Compare

```go
func DoFuncExported() {}

func TestDoFuncExported() {}
```

to

```go
func doFuncNonExported() {}

func Test_doFuncNonExported() {}
```

where the latter is the matching case here.

dhiller replied:
And again - this is just to help the IDE to find the test for the func and the func for the test func.

lyarwood commented:
/cc

lyarwood commented on Makefile:
```diff
@@ -1,6 +1,39 @@
+	go test ./... -coverprofile=/tmp/coverage.out
+	covreport  -i /tmp/coverage.out -o ${COVERAGE_OUTPUT_PATH}
+
+install-metrics-binaries:
```
nit - Any reason to have another target with an unrelated name for installing a linting tool?

dhiller replied:
probably over-engineering :)

lyarwood commented on Makefile:
```diff
@@ -1,6 +1,39 @@
+	covreport  -i /tmp/coverage.out -o ${COVERAGE_OUTPUT_PATH}
+
+install-metrics-binaries:
+	if ! command -V golangci-lint; then curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b ${GOPATH}/bin ${GOLANGCI_LINT_VERSION} ; fi
```
nit - It would be nicer to install this into a local bin to avoid changing the users version

https://github.com/kubevirt/common-instancetypes/blob/cef52ded0abde7b37564faa619aa745d10b5e335/Makefile#L151-L154

lyarwood commented on .golangci.yaml:
```diff
@@ -0,0 +1,6 @@
+  - gocyclo
+linters-settings:
+  gocyclo:
+    min-complexity: 10
```
Missing new line

dhiller commented:
@lyarwood thank you for the review and apologies for the long wait time. I have addressed your comments, but since a rebase happened some linting errors had entered the code base again. I have fixed these now.

PTAL, thank you! :pray:

lyarwood reviewed (approved):
/lgtm

dhiller commented:
:ping_pong: @aburdenthehand :slightly_smiling_face:

aburdenthehand commented:
@xpivarc - you happy?

dhiller commented:
:ping_pong:

aburdenthehand commented:
@dhiller Did you want to squash these commits?
@xpivarc Were your concerns alleviated?

dhiller commented:
:ping_pong: @jean-edouard if you have time :pray:

dhiller commented:
@xpivarc @aburdenthehand @lyarwood would you mind taking another look? Thank you!

dhiller commented:
@aburdenthehand this obviously doesn't scale. I am thinking about moving the code towards project-infra or creating another repository so we can work on improving the code without waiting for reviews from people busy with other obligations.

WDYT?

EdDev commented on Makefile:
```diff
@@ -1,7 +1,21 @@
+.PHONY: generate validate-sigs install-metrics-binaries lint
+
+export GOLANGCI_LINT_VERSION := v1.62.2
```
This is expected to change all the time, therefore, I would recommend not to place it in the makefile.

EdDev replied:
BTW, v2 is already out and is not backward compatible with the configuration. Perhaps it is worth bumping now already.

EdDev commented on validators/cmd/sigs/sigs-validator_test.go:
```diff
@@ -0,0 +1,462 @@
+ *
+ */
+
+package main
```
It is highly recommended to place test files in their own (test) package in order to avoid testing internal implementations.

dhiller replied:
These tests explicitly test an internal implementation, as it's essential. I believe this to be a valuable tradeoff.

WDYT?

EdDev commented on validators/cmd/sigs/testdata/existing-dir/.gitkeep:
What is this file for?

EdDev commented on validators/cmd/sigs/testdata/existing-file:
What is this file for?

dhiller replied:
File required for unit tests.

dhiller commented:
@EdDev thanks for the review, the other PR with job definitions is to be updated soon. I addressed your comments, PTAL! :pray:

EdDev reviewed (approved):
Thanks!

A lot has changed, hope I did not miss anything important.

dhiller commented:
@aburdenthehand seems to be ok now, would you have time to take a quick look please?
